### PR TITLE
[flake8-comprehensions] Account for list and set comprehensions in `unnecessary-literal-within-tuple-call` (`C409`)

### LINKED ISSUE #12648: flake8-comprehensions (C4) - Remove unnecessary list comprehensions in additional contexts that can take arbitrary iterables (C419)
The following code all creates unnecessary list comprehensions

```python
    a = all([i is not None for i in range(10)])
    b = tuple([i * 2 for i in range(10)])
    c = sum([i * 2 for i in range(10)])
    d = ",".join([str(i * 2) for i in range(10)])
```

ruff will only suggest changes on the first line:

```
stuff.py:3:13: C419 Unnecessary list comprehension
  |
2 | def main() -> None:
3 |     a = all([i is not None for i in range(10)])
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C419
4 |     b = tuple([i * 2 for i in range(10)])
5 |     c = sum([i * 2 for i in range(10)])
  |
  = help: Remove unnecessary list comprehension
```

However all of these are unnecessary list comprehensions. It ought to be able to identify them all. For one, the [docs suggest](https://docs.astral.sh/ruff/rules/unnecessary-comprehension-in-call/) that `sum` is covered, but from the above, it appears to not fully be.

### PR DESCRIPTION
**Pull request by dylwil3**
## Summary

Make it a violation of `C409` to call `tuple` with a list or set comprehension, and
implement the (unsafe) fix of calling the `tuple` with the underlying generator instead.

Closes #12648.

## Test Plan

Test fixture updated, cargo test, docs checked for updated description.

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_comprehensions/snapshots/ruff_linter__rules__flake8_comprehensions__tests__C409_C409.py.snap [hunk 1]
```diff
@@ -239,3 +244,146 @@ C409.py:26:6: C409 [*] Unnecessary `list` literal passed to `tuple()` (rewrite a
...
+28    |-tuple([x for x in range(5)])
+   28 |+tuple(x for x in range(5))
+29 29 | tuple({x for x in range(10)})
+30 30 | tuple(x for x in range(5))
+31 31 | tuple([
+
+C409.py:29:1: C409 [*] Unnecessary `set comprehension` literal passed to `tuple()` (remove the outer call to `tuple()`)
+   |
+28 | tuple([x for x in range(5)])
+29 | tuple({x for x in range(10)})
```
Comment (Skylion007) [inline crates/ruff_linter/src/rules/flake8_comprehensions/snapshots/ruff_linter__rules__flake8_comprehensions__tests__C409_C409.py.snap:275]: This is not a safe transform, a set could be used here for deduplicating elements in the iterable before the tuple is created. I'm pretty sure this is only safe / sensible to flag for list comprehensions.

**Comment by Skylion007** [comment]
I'd argue this is only safe to flag for list comprehensions as set comprehensions will dedup and potentially reorder elements as well (not that you should rely on that nondeterm ordering but still).

**Comment by dylwil3** [comment]
Good point! Another thing I noticed is that the suggestion text doesn't make sense in the context of comprehensions. I'll fix both next time I'm in front of a computer.

Thanks!

**Comment by dylwil3** [comment]
Ok, should be good to go!

**Review by charliermarsh — APPROVED** [review]
Thanks! (FYI I had to move this behavior to preview to adhere to our versioning policy since it's an expansion in scope.)

**Comment by hauntsaninja** [comment]
I find this kind of thing unfortunate, because `tuple(i for i in range(10000))` is like 50% slower than `tuple([i for i in range(10000)])`.

There's no way to configure ruff to disambiguate between the two.
This is a similar complaint to https://github.com/astral-sh/ruff/issues/8884

**Pull request merged**